### ISSUE TITLE
Sidecar envoy proxy readiness probe failing - Istio 1.6.7

### ISSUE THREAD

**Issue by Siddharthk** [istio/istio#26792]
Bug description

Sidecar envoy proxy is failing readiness probe intermittently in istio 1.6.7. Is there a way we can increase readiness probe timeout for the side car proxy?

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Expected behavior

Sidecar envoy proxy should not be failing.

Steps to reproduce the bug:
Running an app with istio side car injected:
```
dev         app-amaikei6-6d6dc5c474-2sl84       1/2     Running            0          2d19h
```
Logs below:
```
2020-08-21T17:33:30.669986Z     info    Envoy proxy is ready
2020-08-21T18:01:58.066886Z     info    Subchannel Connectivity change to CONNECTING
```
After 2-3 days:
```
2020-08-24T08:16:09.602127Z     info    Channel Connectivity change to READY
2020-08-24T08:21:15.651523Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:17.353903Z     info    Envoy proxy is ready
2020-08-24T08:21:27.694610Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:29.790957Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:31.823298Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-08-24T08:21:33.989034Z     warn    Envoy proxy is NOT ready: failed to get readiness stats: Get "http://127.0.0.1:15000/stats?usedonly&filter=^(server.state|listener_manager.workers_started)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2020-08-21T17:33:23Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2020-08-24T08:22:25Z"
    **message: 'containers with unready status: [istio-proxy]'**
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2020-08-24T08:22:25Z"
    message: 'containers with unready status: [istio-proxy]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2020-08-21T17:33:21Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://9b6f57b03ee4347912b1f75593308c012b7170285c5e8c0e128c298b37fc636a
    image: istio/proxyv2:1.6.7
    imageID: docker-pullable://istio/proxyv2@sha256:885765eea98e0123e2545df3e708257a651fdc31a7b512069f392967d015c1fc
    lastState: {}
    name: istio-proxy
    ready: false
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2020-08-21T17:33:28Z"

```
Version (include the output of istioctl version --remote and kubectl version and helm version if you used Helm)

client version: 1.6.7
control plane version: 1.6.7
data plane version: 1.6.7 (78 proxies)

Client Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.8-eks-e16311", GitCommit:"e163110a04dcb2f39c3325af96d019b4925419eb", GitTreeState:"clean", BuildDate:"2020-03-27T22:40:13Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.8-eks-e16311", GitCommit:"e163110a04dcb2f39c3325af96d019b4925419eb", GitTreeState:"clean", BuildDate:"2020-03-27T22:37:12Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}

How was Istio installed?

Using istioctl 

Environment where bug was observed (cloud vendor, OS, etc)

AWK EKS 1.16.8

**Comment by ceastman-ibm** [comment]
I see the same behavior in istio 1.6.8

**Comment by GregHanson** [comment]
I think this is a potential bug.  There is an env var for determining the timeout `pilot-agent` uses when calling Envoy's `/stats` API, `ENVOY_READINESS_CHECK_TIMEOUT`, and the default value it 5 seconds ([here](https://github.com/istio/istio/blob/master/pilot/cmd/pilot-agent/status/util/stats.go#L41)).  But this is essentially useless since the readiness probe has a timeout of 1 second. Ingress gateway example [here](https://github.com/istio/istio/blob/master/manifests/charts/gateways/istio-ingress/templates/deployment.yaml#L147
).

Seems related to this discuss.io topic as well: https://discuss.istio.io/t/http-127-0-0-1-15000-stats-usedonly-response-takes-a-long-time/7241

**Comment by GregHanson** [comment]
Also, istio doesn't explicitly define a istio-proxy readiness probe timeout value - and so it is falling back to the Kubernetes default:
https://github.com/istio/istio/blob/master/manifests/charts/istio-control/istio-discovery/files/injection-template.yaml#L286-L292
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes

**Comment by linsun** [comment]
Agree with @GregHanson - Seems our ingress gw readiness check call is only 1s and a bit too short. @ceastman-ibm has indicated that it takes 2-3s sometimes on his ingress pod.

Does anyone have any issue if we increase this to 5s which matches envoy's default?  cc @howardjohn @lizan @PiotrSikora @costinm @nrjpoddar @rshriram

**Comment by linsun** [comment]
Also, should we add startup Probe or liveness Probe so that if the check takes too long, the pod can be auto restarted?  Reading from k8s doc, the restart policy (default to always) only works on `livenessProbe` or `startupProbe` - https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/

**Comment by howardjohn** [comment]
I don't think we need any of the above if I am understanding right, we have already fixed this in master: https://github.com/istio/istio/pull/26820

**Comment by GregHanson** [comment]
Is there any scenario where we don't want `ENVOY_READINESS_CHECK_TIMEOUT` to be in sync with the readiness probe timeout value?  Don't we at least want them to have the same default value?

**Comment by ramaraochavali** [comment]
Yes. I think it would be good have them in sync.

**Comment by linsun** [comment]
Agreed, while #26820 minimizes the changes of hitting it, having the value in sync would be good and reduces confusion as well :)

**Comment by howardjohn** [comment]
This is fixed in 1.7+ by https://github.com/istio/istio/commit/af1f055004f20340b3c3d1fec8b0d9434e2fa892
